### ISSUE TITLE
How to specify class weights in model training?

### ISSUE THREAD

**Issue by mucun1988** [UKPLab/sentence-transformers#2416]
I am having a very imbalanced training dataset. Is there a way I could specify class weights (e.g., class 0: 0.1, class 1: 1) for cross encoder training?

**Comment by ir2718** [comment]
@mucun1988 

If you're using the `SoftmaxLoss`, you can achieve this by defining the weight in the underlying `CrossEntropyLoss` object. Here's an example for a dataset containing three labels:

```
num_labels = 3
label_counts = {}
for k in range(num_labels):
    label_counts[k] = [x.label for x in train_samples if x.label == k][0]

class_weights = torch.tensor([sum(label_counts.values())/v for k, v in label_counts.items()])

loss = losses.SoftmaxLoss(
    model=model,
    sentence_embedding_dimension=model.get_sentence_embedding_dimension(),
    num_labels=3,
    loss_fct=nn.CrossEntropyLoss(weight=class_weights),
)
```

**Comment by mucun1988** [comment]
Many thanks @ir2718! 

Taking this one as an example
https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/cross-encoder/training_nli.py#L73

I should do

`model.fit(
    train_dataloader=train_dataloader,
    evaluator=evaluator,
    epochs=num_epochs,
    evaluation_steps=10000,
    warmup_steps=warmup_steps,
    output_path=model_save_path,
   loss_fct= loss)`?

**Comment by tomaarsen** [comment]
Hello!

@ir2718 their example should do well with a SentenceTransformer. If you're training a CrossEncoder like your example, then I think you can pass a CrossEntropyLoss instance directly:
```python
num_labels = 3
label_counts = {}
for k in range(num_labels):
    label_counts[k] = len([x.label for x in train_samples if x.label == k])

class_weights = torch.tensor([sum(label_counts.values())/v for v in label_counts.values()])
loss = nn.CrossEntropyLoss(weight=class_weights)

# Train the model
model.fit(
    train_dataloader=train_dataloader,
    loss_fct=loss,
    evaluator=evaluator,
    epochs=num_epochs,
    evaluation_steps=10000,
    warmup_steps=warmup_steps,
    output_path=model_save_path,
)
```

- Tom Aarsen

**Comment by ir2718** [comment]
@mucun1988 

I overlooked the fact that you're using a CrossEncoder, while the code I provided is suited for a SentenceTransformer. Tom's example should work and it takes care of calling the `fit` method as well.

**Comment by mucun1988** [comment]
ty @tomaarsen @ir2718! this is exactly what I am looking for!
